Fix "Please adopt containerBackgroundAPI" for iOS 17 Widgets

staskus opened the pull request:
Partially addresses #21672

The goal of this PR is to conform to the new iOS 17 Widget APIs to avoid "Please adopt containerBackgroundAPI" message:
- Implement an empty `containerBackground` for conformance. This API does not support the usage of `AccessoryWidgetBackground`, therefore I decided to remove the background altogether to fit the current design guidelines
- Update UI of HomeScreenWidget to support removable background
- Temporarily disabled iOS 17 margins to support StandBy mode

I want to make even better iOS 17 support in the short-term future. Plus, we probably should make naming changes, given HomeScreen widgets work on iPadOS Lock Screen as well.

## To test:

- Lock Screen and Home screen widgets should work as before
- Home Screen widgets and Lock Screen Widgets should now be available on iPadOS Lock Screen vertical and horizontal versions

![image](https://github.com/wordpress-mobile/WordPress-iOS/assets/4062343/198758d2-c9f3-4798-b025-807bd964050f)

## Regression Notes
1. Potential unintended areas of impact

Should be none, Widgets should work and appear as before

2. What I did to test those areas of impact (or what existing automated tests I relied on)

Manual testing

3. What automated tests I added (or what prevented me from doing so)

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

UI Changes testing checklist:
- [ ] Portrait and landscape orientations.
- [ ] Light and dark modes.
- [ ] Fonts: Larger, smaller and bold text.
- [ ] High contrast.
- [ ] VoiceOver.
- [ ] Languages with large words or with letters/accents not frequently used in English.
- [ ] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [ ] iPhone and iPad. 
- [ ] Multi-tasking: Split view and Slide over. (iPad)

staskus committed:
Disable additional iOS17 margins for widgets

staskus committed:
Conform to iOS 17 removable background API on Lock Screen Widgets

Conforming by implementing containerBackground API.

For now, Lock Screen Widgets keep using AccessoryWidgetBackground(), since this background is not supported by removable background.

staskus committed:
Disable background removing for Home Screen widgets

Home Screen widgets require layout adjustment to support background removal

staskus commented:
@guarani 

This is not a high priority but I wanted to at least conform to these APIs so we wouldn't need to make quick beta fixes when CI is updated to Xcode 15.

staskus commented:
@guarani review is not necessary right now.

CI throws an issue:

> value of type 'some WidgetConfiguration' has no member 'contentMarginsDisabled'

It means that we cannot make these changes until Xcode is updated on CI, so we cannot merge it right now.

staskus committed:
Silence Xcode 14 warnings by wrapping new API calls in preprocessor macros

staskus committed:
Update iOS17WidgetAPIs.swift

staskus commented on WordPress/JetpackStatsWidgets/Widgets/HomeWidgetAllTime.swift:
```diff
@@ -33,5 +33,7 @@ struct HomeWidgetAllTime: Widget {
         .description(LocalizableStrings.allTimePreviewDescription)
         .supportedFamilies([.systemSmall, .systemMedium])
+        .iOS17ContentMarginsDisabled() /// Temporarily disable additional iOS17 margins for widgets
+        .iOS17ContainerBackgroundRemovable(false) /// Temporarily disable iOS 17 background removing functionality
```
It doesn't work

guarani reviewed (commented):
To summarize what we discussed in Slack, p1697208501951699/1697187055.060939-slack-C441E3YTS, once`containerBackground` was implemented, Lock Screen widgets no longer showed the `Please adopt containerBackground API`, but Home Screen widgets still do, despite implementing `containerBackgroundRemovable` and passing in `false`.

staskus committed:
Change Home Screen Small Widget design if it's displayed without background

Small Home Screen Widget can be shown in iPad lock screen on horizontal orientation:
- Change the order of title and value
- Increase font size of value

staskus commented:
## Home Widgets (System Small)

I added to support background removal APIs to Small Home Screen Widgets. For them to appear nicely, I changed the order of value title and value and made the value font larger. They can be added on the iPad Lock Screen with a Horizontal Orientation.

These are the 2 variations that system small widgets can be presented in:

<img width="360" alt="Small (Lock Screen)" src="https://github.com/wordpress-mobile/WordPress-iOS/assets/4062343/16314776-270a-4539-a551-ce77e9a5a414">
<img width="302" alt="Small (Home Screen)" src="https://github.com/wordpress-mobile/WordPress-iOS/assets/4062343/1d953885-dbf9-4f4d-abf3-ec11d524b2d6">

The inspiration was a weather widget example:

<img width="202" alt="weather 2" src="https://github.com/wordpress-mobile/WordPress-iOS/assets/4062343/9a562ce5-6b09-467b-8b28-6e309cc44233">
<img width="221" alt="Weather Beforre" src="https://github.com/wordpress-mobile/WordPress-iOS/assets/4062343/2fa769c5-b32f-45cb-b557-42df3708fa26">

## Lock Screen Widgets (Accessory Rectangular)

To make the selection consistent with iOS 17 design guidelines and newest widgets, I think the best option is to remove the background altogether:

![IMG_1040](https://github.com/wordpress-mobile/WordPress-iOS/assets/4062343/0237827f-8391-4992-aabd-0a70d6cc68bd)

Now all these widgets can be displayed together on the iPad Lock Screen:

![Simulator Screenshot - iPad Air (5th generation) - 2023-10-18 at 11 34 01](https://github.com/wordpress-mobile/WordPress-iOS/assets/4062343/2eb0321d-09d0-4646-8076-9c70fc878e7d)

staskus commented:
@guarani 

My plan is as follows:

1. Get feedback from designers about widgets without background pc8eDl-14z-p2#comment-935
2. Split this PR into two: 1) Widget Lock Screen Widget UI updates without a background and 2) iOS 17 APIs conformance
3. Review, test & merge **PR 1)** since it would not require Xcode changes, mostly it would remove background from lock screen widgets
4. Review, test & merge https://github.com/wordpress-mobile/WordPress-iOS/pull/21514, since it would be counter-productive to test it while this PR is still ongoing
5. Ideally, we would ship lock screen widget background updates and this week widget with 23.6
6. Once we enable Xcode 15, merge **2) iOS 17 APIs conformance** PR.

staskus committed:
Remove iOS17ContainerBackgroundRemovable

staskus committed:
Set removableWidgetBackground for unimplemented text

staskus committed:
Update Home Screen widgets according to design requirements

staskus committed:
Add icon to title view

staskus committed:
Update label color of unconfigured view

guarani commented:
FYI: The PR that updates CI to Xcode 15 is targeting Monday's 24.0 release: https://github.com/wordpress-mobile/WordPress-iOS/pull/21921

We could spend some time this week to get necessary widget changes done for Monday's beta cut. Wdyt? I haven't been focusing on these PRs but I think it makes sense to do so now.

staskus commented:
> FYI: The PR that updates CI to Xcode 15 is targeting Monday's 24.0 release: https://github.com/wordpress-mobile/WordPress-iOS/pull/21921

@guarani, thanks for making a note! Yes, I followed it but I wasn't sure if it would make it on time for 24.0. 

> Wdyt? I haven't been focusing on these PRs but I think it makes sense to do so now.

Yes, once they merge Xcode 15 updates we'll need to get this PR merged for 24.0 as well. It would be good to build it with Xcode 15 locally and review it so it would be ready. 👍

mokagio reviewed (commented):
I'll start by saying that I haven't looked at the whole code, so apologies if the question has an easy answer. But I wonder if any of this could go into [`JetpackStatsWidgetsCore`](https://github.com/wordpress-mobile/WordPress-iOS/tree/4e9bb83c82264e642cdd36e4456ff1963577d3a7/Modules/Sources/JetpackStatsWidgetsCore).

staskus commented:
> I'll start by saying that I haven't looked at the whole code, so apologies if the question has an easy answer. But I wonder if any of this could go into [`JetpackStatsWidgetsCore`](https://github.com/wordpress-mobile/WordPress-iOS/tree/4e9bb83c82264e642cdd36e4456ff1963577d3a7/Modules/Sources/JetpackStatsWidgetsCore).

Great question, @mokagio. I'll check! I made this PR 3 months ago to prepare for Xcode 15 updates when this module didn't exist.

staskus committed:
Merge branch 'trunk' into fix/please-adopt-container-background-api-for-widgets-xcode-15-ios-17

staskus committed:
Code structure

staskus commented:
> I'll start by saying that I haven't looked at the whole code, so apologies if the question has an easy answer. But I wonder if any of this could go into [JetpackStatsWidgetsCore](https://github.com/wordpress-mobile/WordPress-iOS/tree/4e9bb83c82264e642cdd36e4456ff1963577d3a7/Modules/Sources/JetpackStatsWidgetsCore).

I checked the affected files and I don't think any of them needs to go to `JetpackStatsWidgetCore`. Changes are only to Stats Widget UI code which can continue belonging to `JetpackStatsWidgets`. `JetpackStatsWidgetCore` should contain data types and functionality that needs to be reused between targets.

mokagio commented on WordPress/JetpackStatsWidgets/Views/SingleStatView.swift:
```diff
@@ -1,19 +1,59 @@
-    let viewData: GroupedViewData
+    @Environment(\.showsWidgetContainerBackground) var showsWidgetContainerBackground: Bool
+
+    init(viewData: GroupedViewData) {
+        self.title = viewData.widgetTitle
+        self.description = viewData.siteName
+        self.valueTitle = viewData.upperLeftTitle
+        self.value = viewData.upperLeftValue
+    }
```
Did you consider delegating to `self.init(title:...`?

staskus replied:
Good observation. I'll use `self.init..`

mokagio commented on WordPress/JetpackStatsWidgets/LockScreenWidgets/LockScreenStatsWidget.swift:
```diff
@@ -33,5 +33,6 @@ struct LockScreenStatsWidget<T: LockScreenStatsWidgetConfig>: Widget {
         .configurationDisplayName(config.displayName)
         .description(config.description)
         .supportedFamilies(config.supportFamilies)
+        .iOS17ContentMarginsDisabled() /// Temporarily disable additional iOS17 margins for widgets
```
**Super nitpick**, `///` is a documentation comment, as opposed to the standard `//`. With `///` Xcode and other tools will render the comment as documentation reading tags and markdown. However, the comment needs to be on a type or method declaration. We don't need to "document" method calls. I just wanted to point it out but I don't expect this to be addressed. 👍 😄

mokagio committed:
Merge remote trunk into fix/please-adopt-container-background-api-for-widgets-xcode-15-ios-17

This is necessary to get up to date with a CI configuration change that
requires running on Xcode 15.1.

See https://github.com/wordpress-mobile/WordPress-iOS/pull/22270

mokagio commented on WordPress/JetpackStatsWidgets/Views/Cards/StatsValueView.swift:
```diff
@@ -30,5 +30,6 @@ struct StatsValueView: View {
             .fontWeight(fontWeight)
             .foregroundColor(foregroundColor)
             .lineLimit(lineLimit)
+            .minimumScaleFactor(0.5)
```
Why 0.5?

staskus replied:
0.5 is an arbitrary value, could be smaller or a bit larget as well.

Previously created rectangular widgets can now appear without a background on iPadOS Horizontal Lock Screen. Since they appear without a background, I increased font sizes to have less space and resemble the look of other widgets without a background.

However, when numbers become large, I want font to Value font to shrink, to be fully displayed.

Without `minimumScaleFactor`:
<img width="265" alt="image" src="https://github.com/wordpress-mobile/WordPress-iOS/assets/4062343/729028e3-63b5-4070-b0d0-69def7f192b6">

With `minimumScaleFactor`:
<img width="200" alt="image" src="https://github.com/wordpress-mobile/WordPress-iOS/assets/4062343/c1de6449-238b-4407-a585-d617d11a2ef5">

mokagio commented:
@staskus thank you for looking into whether some files could go into the module.

> `JetpackStatsWidgetCore` should contain data types and functionality that needs to be reused between targets.

Yes, and also any kind of standalone logic that we might want (ought to!) unit test. We can't run unit tests on app extension targets, so anything we want to test should go into that Swift Package. 

As far as I can see, this PR is all about SwiftUI configurations so there's nothing that fits that criteria either 👍

mokagio commented on WordPress/JetpackStatsWidgets/LockScreenWidgets/Views/LockScreenMultiStatView.swift:
```diff
@@ -31,9 +31,11 @@ struct LockScreenMultiStatView: View {
             .accessibilityElement(children: .combine)
         } else {
             Text("Not implemented for widget family \(family.debugDescription)")
+                .removableWidgetBackground()
```
Just curious about the Swift UI architecture, which is not my strong suit. Would it have been possible to group the conditional logic, e.g. in an `*Stack`, to only write the `removableWidgetBackground()` transformation once?

staskus replied:
Good question. I think it would be possible 👍

mokagio reviewed (approved):
Run it on the Simulator with Xcode 15.1 after merging `trunk` into this branch. Everything looks good as far as I can see, but I have little context around the expected widgets look and feel.

Let's merge this so it's part of 24.0, get feedback from @osullivanchris (Chris has already commented on the widgets, see pc8eDl-14z-p2), and address possible tweaks as beta fixes.

### iPhone

<table>
<tr>
<td>
<img width="559" alt="Screenshot 2024-01-08 at 16 37 04" src="https://github.com/wordpress-mobile/WordPress-iOS/assets/1218433/fa73ff0f-efb7-4a43-89ee-bbde504d0926">
</td>
<td>
<img width="426" alt="image" src="https://github.com/wordpress-mobile/WordPress-iOS/assets/1218433/8157f64f-1e22-421d-abaa-36cdce1e9470">
</td>
<td>
<img width="559" alt="Screenshot 2024-01-08 at 16 43 45" src="https://github.com/wordpress-mobile/WordPress-iOS/assets/1218433/b6a88e65-ac88-41ab-8763-f12ed464609c">
</td>
</tr>
</table>

### iPad

<table>
<tr>
<td>
<img width="858" alt="Screenshot 2024-01-08 at 16 52 40" src="https://github.com/wordpress-mobile/WordPress-iOS/assets/1218433/c30907ca-6652-4122-bf0e-83eb6689d8b6">

</td>
<td>

<img width="858" alt="Screenshot 2024-01-08 at 16 53 38" src="https://github.com/wordpress-mobile/WordPress-iOS/assets/1218433/9c3277f7-1ae8-4392-8b27-e80b0e9a342e">


</td>
</tr>
</table>

staskus commented:
Thanks for the review, @mokagio!

I can merge it to 24.0. 

This PR was all about basic iOS 17 support. It meant supporting `removableWidgetBackground` and also making sure existing medium and large widgets could appear without a background on the iPad OS Horizontal Lock Screen.

I agree we can make other tweaks after we merge this PR.

mokagio commented:
I think the fact the latest build on this branch was cancelled made all the CI checks go "Errored"

![image](https://github.com/wordpress-mobile/WordPress-iOS/assets/1218433/e5ed2343-86e4-41a4-aea8-ec7502861c08)

I started a new one so we can merge this ASAP.

@staskus what do you mean with "I can merge it to 24.0." and with the "[Status] Blocked" label? I read this as "there's a reason why we shouldn't merge it to `trunk` right now" but I don't see what the reason is 🤔

staskus commented:
> I started a new one so we can merge this ASAP.

@mokagio 

Okay. Sorry, there was a bit of a misunderstanding. I missed the fact you already merged it with trunk. Yes, we can merge it whenever CI succeeds. 👍

guarani commented:
I forgot to review on Friday 🙈  Thanks for handling this folks!
